### PR TITLE
[build] [wip] Support for the dune build system.

### PR DESCRIPTION
**Pull request by ejgallego**
This is a preliminary draft PR for building `zarith` with dune; my motivation was to be able to compose `zarith` with other developments, so far seems to work well!

If developers like this, I could finish the TODO:

- configuration
- fix warnings
- run all tests
- opam generation?
- CI?

### CODE REVIEW

**Comment by olafhering** [comment]
This needs rebasing. `fact`, `fib` and `tak` do not exist anymore.

**Comment by hannesm** [comment]
is this planned to be rebased & merged & zarith to be released (with the very nice other changes that are so far only in trunk)? thanks.

**Comment by ejgallego** [comment]
I'd be happy to do more progress on this, once I get feedback from the developers as whether this would be considered.

**Comment by antoinemine** [comment]
I have no strong opinion. I think that the main complexity of the build system for Zarith lies in the configure script. The Makefile is rather simple, we have few files and directories.
To get a more decisive opinion and be excited about this, it would help to see a complete dune-based build system with the configure phase working on all the platforms currently supported by the Makefile, as well as some incentive (what is the concrete benefit of switching to dune ?).

**Comment by xavierleroy** [comment]
My Cryptokit library recently moved to using Dune: https://github.com/xavierleroy/cryptokit/pull/24 .  On the one hand, the previous build system (based on Oasis) had issues, so Dune was an improvement.  On the other hand, Dune's support for autoconfiguration is rather primitive, to the point that I'd rather stick with a handwritten `configure` shell script.

**Comment by ejgallego** [comment]
> To get a more decisive opinion and be excited about this, it would help to see a complete dune-based build system with the configure phase working on all the platforms currently supported by the Makefile, as well as some incentive (what is the concrete benefit of switching to dune ?).

My main motivation to write this was to be able to debug `zarith` and a client [in this case Coq] at the same time, with Dune you only need to drop `zarith` inside Coq's build scope and things work out of the box. This was very useful but don't expect to have a need for it again soon.

Dune also provides some other advantages and extra tooling such as documentation, dune-release, etc... YMMV.

> On the other hand, Dune's support for autoconfiguration is rather primitive, to the point that I'd rather stick with a handwritten configure shell script.

I don't think that the configure script would be replaced, only adapted to output files that can be `:include`'d from dune's build rules.

**Comment by EduardoRFS** [comment]
I would also like to have it using dune + autoconf, like opam itself does it. It simplifies a lot when cross compiling zarith to other platforms.

File: dune [hunk 1]
```diff
@@ -0,0 +1,22 @@
+(library
+ (name zarith)
+ (wrapped false)
+ (flags -w -6)
+ (library_flags -cclib -lgmp)
```
Comment (kit-ty-kate) [inline dune:5]: I would suggest to use `dune-configurator` here instead, having `-lgmp` as a fallback.
See https://github.com/ocaml/opam-repository/issues/18131

**Comment by ejgallego** [comment]
Great @hhugo , please feel free to take over / close this PR, it is unlikely I can put any cycles on this any time soon.

**Comment by hhugo** [comment]
> I have no strong opinion. I think that the main complexity of the build system for Zarith lies in the configure script. The Makefile is rather simple, we have few files and directories.
> To get a more decisive opinion and be excited about this, it would help to see a complete dune-based build system with the configure phase working on all the platforms currently supported by the Makefile, as well as some incentive (what is the concrete benefit of switching to dune ?).

I've updated https://github.com/hhugo/zarith/tree/dune to demonstrate what a near complete dune-based build system would look like for zarith.

**Comment by ejgallego** [comment]
Thanks a lot @hhugo , I'm closing this PR in favor of #143

**Pull request closed without merging**